src/test_sudoku.c

```diff
@@ -0,0 +1,134 @@
+/**
+ * @file test_sudoku.c
+ *
+ * Purpose  : This function tests whether the regenerated sudoku meets all the sudoku rules or not .
+ *
+ * Language : C
+ *
+ * Author   : Rishith Rao - rishithrao@cmail.carleton.ca
+ *            Prasanth Kota - prasanthkota@cmail.carleton.ca
+ *
+ * Date     : 13 June, 2020.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "../include/sudoku.h"
+
+struct node {
+    int data;
+    struct node *next;
+};
+
+struct node *head = NULL;
+struct node *current = NULL;
+
+/**
+ * Deletes the first element of the array.
+ */
+struct node* deleteFirst() {
+    struct node *tempLink = head;/*save reference to first link*/
+    head = head->next;/*mark next to first link as first*/
+    return tempLink;/*return the deleted link*/
+}
+
+/**
+ * Checks whether the list is empty or not.
+ */
+bool isEmpty() {
+    return head == NULL;
+}
+
+/**
+ * Insert link at the first location.
+ */
+void insertFirst(int data){
+    struct node *link = (struct node*) malloc(sizeof(struct node));/*create a link*/
+    link->data = data;
+    link->next = head;/*point it to old first node*/
+    head = link;/*point first to new first node*/
```

Comment:
Why are you using linked list? Linked list are used when you need to rearrange the data in a certain order. The sudoku_tester() function should be about comparing elements to determine if two digits repeat themselves. I don't see the use of linked list when you can easily just create a copy of 2D array and simply create functions to verify the rows, cols and 3x3 blocks.

Reply:
There is a use of linked lists here. We are copying the values from 1 to 9 to the linked lists and checking the elements of rows/columns/3x3 have values from 1 to 9 or not. So, the linked list will be left with the values that has to be appeared in the place of the repeated values in the row/column/3x3 blocks. This is also modified to simple function for better understanding for the other developers.